src/themis/secure_cell.c

```diff
@@ -101,6 +102,121 @@ themis_status_t themis_secure_cell_decrypt_seal(const uint8_t* master_key,
+                                                                size_t passphrase_length,
+                                                                const uint8_t* user_context,
+                                                                size_t user_context_length,
+                                                                const uint8_t* message,
+                                                                size_t message_length,
+                                                                uint8_t* encrypted_message,
+                                                                size_t* encrypted_message_length)
+{
+    themis_status_t res = THEMIS_FAIL;
+    size_t auth_token_length = 0;
+    size_t ciphertext_length = 0;
+    size_t total_length = 0;
+
+    THEMIS_CHECK_PARAM(encrypted_message_length != NULL);
+
+    /*
+     * Since Seal mode prepends authentication token to the message
+     * we need to get the size of this token at first.
+     */
+    res = themis_auth_sym_encrypt_message_with_passphrase(passphrase,
+                                                          passphrase_length,
+                                                          message,
+                                                          message_length,
+                                                          user_context,
+                                                          user_context_length,
+                                                          NULL,
+                                                          &auth_token_length,
+                                                          NULL,
+                                                          &ciphertext_length);
+    if (res != THEMIS_BUFFER_TOO_SMALL) {
+        return res;
+    }
+
+    total_length = auth_token_length + ciphertext_length;
+    if (!encrypted_message || *encrypted_message_length < total_length) {
+        *encrypted_message_length = total_length;
+        return THEMIS_BUFFER_TOO_SMALL;
+    }
+
+    res = themis_auth_sym_encrypt_message_with_passphrase(passphrase,
+                                                          passphrase_length,
+                                                          message,
+                                                          message_length,
+                                                          user_context,
+                                                          user_context_length,
+                                                          encrypted_message,
+                                                          &auth_token_length,
+                                                          encrypted_message + auth_token_length,
+                                                          &ciphertext_length);
+    if (res == THEMIS_SUCCESS || res == THEMIS_BUFFER_TOO_SMALL) {
+        *encrypted_message_length = auth_token_length + ciphertext_length;
```

Comment:
Reuse `total_length` here?

Reply:
This is intentional because the themis_auth_sym_encrypt_message_with_passphrase() call might have modified those values. Of course, it does not modify them and accurately reports required lengths so we could reuse the value. But I'd leave that for future-proofing.

For all functions that return buffers, it's not strictly incorrect behavior to return slightly _bigger_ required buffer size during measurement stage (BUFFER_TOO_SMALL) and then return accurate – and possibly smaller – actual size of the buffer used on SUCCESS. Currently we are able to accurately determine required length without actually processing the data, but it might not be the case in the future.